### PR TITLE
Pizza box additions

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Pizza box stacks can now fall over 
imageadd: Pizza box inhands now stacks depending on how many you're holding.
/:cl:

_"Xhuis," spoke MSO, "the servers are crashing!"
"Xhuis," spoke oranges, "people can reconnect infinitely!"
"Xhuis," spoke the maintainers, "we need you to code something!"_

So I coded **this.**

A list of changes, in no particular order:
* Pizza boxes no longer use icon states to determine stacks. Instead, it uses overlays depending on the stack size.
* By using the system I introduced with greyscaling screwdrivers, pizza box inhands now stack more pizza boxes dynamically depending on how many you're carrying.
* Pizza boxes can now stack, functionally, up to 9. Any boxes past 5 will have an increasingly high chance to topple the stack, and at 10 they will completely collapse regardless.
* Attacking others with a stack or throwing a stack now has a chance to topple the stack.
* Toppling the stack has been added, and will cause every pizza box in the stack to fall in a random direction. Each box has a 50% chance of spilling its pizza, which will also move in a random direction.

### CODE REVIEW

**Comment by Anonmare** [comment]
Speedmerge now

**Comment by Xhuis** [comment]
Point of note: a lot of this code is oldcode and is thus pretty bad. I'll touch up on it tomorrow!

**Comment by Tacolizard** [comment]
for memes, the damage from using this as a weapon should increase the more boxes there are

File: code/modules/food_and_drinks/pizzabox.dm [hunk 1]
```diff
@@ -68,13 +68,30 @@
...
 		var/obj/item/pizzabox/box = boxes.len ? boxes[boxes.len] : src
 		if(box.boxtag != "")
 			var/mutable_appearance/tag_overlay = mutable_appearance(icon, "pizzabox_tag")
 			tag_overlay.pixel_y = boxes.len * 3
 			add_overlay(tag_overlay)
 
+/obj/item/pizzabox/worn_overlays(isinhands, icon_file)
+	. = list()
+	var/current_offset = 2
+	if(isinhands && boxes.len)
```
Comment (Cyberboss) [inline code/modules/food_and_drinks/pizzabox.dm]: No need for the boxes.len check

File: code/modules/food_and_drinks/pizzabox.dm [hunk 2]
```diff
@@ -135,22 +153,29 @@
...
-				to_chat(user, "<span class='notice'>You put [newbox] on top of [src]!</span>")
-				newbox.update_icon()
-				update_icon()
+			if(!user.drop_item())
 				return
-			else
-				to_chat(user, "<span class='notice'>The stack is dangerously high!</span>")
+			boxes += add
+			newbox.boxes.Cut()
+			newbox.loc = src
```
Comment (Cyberboss) [inline code/modules/food_and_drinks/pizzabox.dm]: forceMove

File: code/modules/food_and_drinks/pizzabox.dm [hunk 3]
```diff
@@ -135,22 +153,29 @@
...
+					disperse_pizzas()
+				else
+					to_chat(user, "<span class='warning'>The stack is getting a little high...</span>")
+			return
 		else
 			to_chat(user, "<span class='notice'>Close [open ? src : newbox] first!</span>")
 	else if(istype(I, /obj/item/weapon/reagent_containers/food/snacks/pizza) || istype(I, /obj/item/weapon/reagent_containers/food/snacks/customizable/pizza))
 		if(open)
+			if(pizza)
+				to_chat(user, "<span class='warning'>[src] already has a [pizza.name]!</span>")
```
Comment (Cyberboss) [inline code/modules/food_and_drinks/pizzabox.dm]: use \a [pizza]

File: code/modules/food_and_drinks/pizzabox.dm [hunk 4]
```diff
@@ -200,6 +225,34 @@
...
+	visible_message("<span class='warning'>The pizzas fall everywhere!</span>")
+	for(var/V in boxes)
+		var/obj/item/pizzabox/P = V
+		var/fall_dir = pick(GLOB.alldirs)
+		step(P, fall_dir)
+		if(P.pizza && prob(50)) //rip pizza
+			P.open = TRUE
+			P.pizza.forceMove(get_turf(P))
+			fall_dir = pick(GLOB.alldirs)
+			step(P.pizza, fall_dir)
```
Comment (Cyberboss) [inline code/modules/food_and_drinks/pizzabox.dm]: Should check for isdenseturf

File: code/modules/food_and_drinks/pizzabox.dm [hunk 5]
```diff
@@ -200,6 +225,34 @@
...
+/obj/item/pizzabox/throw_impact(atom/movable/AM)
+	if(boxes.len >= 2 && prob(20 * boxes.len))
+		disperse_pizzas()
+
+/obj/item/pizzabox/proc/disperse_pizzas()
+	visible_message("<span class='warning'>The pizzas fall everywhere!</span>")
+	for(var/V in boxes)
+		var/obj/item/pizzabox/P = V
+		var/fall_dir = pick(GLOB.alldirs)
+		step(P, fall_dir)
```
Comment (Cyberboss) [inline code/modules/food_and_drinks/pizzabox.dm]: Should check for isdenseturf

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/food_and_drinks/pizzabox.dm (hunk 1 above)
Comment (RemieRichards) [inline code/modules/food_and_drinks/pizzabox.dm]: just before you ask, for each loops don't run if the list is empty, so the check is unnecessary.

File: code/modules/food_and_drinks/pizzabox.dm (hunk 5 above)
Comment (RemieRichards) [inline code/modules/food_and_drinks/pizzabox.dm]: no need, step doesn't go through walls, the box just won't move if it tries to step onto a wall/something dense.

File: code/modules/food_and_drinks/pizzabox.dm (hunk 4 above)
Comment (RemieRichards) [inline code/modules/food_and_drinks/pizzabox.dm]: see above.

**Comment by Hatfish** [comment]
_If I flip the pizzas, Mr. Aziz will flip out!_

File: code/modules/food_and_drinks/pizzabox.dm [hunk 6]
```diff
@@ -135,22 +153,29 @@
...
-			if((boxes.len + 1) + add.len <= 5)
-				if(!user.drop_item())
-					return
-				boxes += add
-				newbox.boxes.Cut()
-				newbox.loc = src
-				to_chat(user, "<span class='notice'>You put [newbox] on top of [src]!</span>")
-				newbox.update_icon()
-				update_icon()
+			if(!user.drop_item())
```
Comment (Cyberboss) [inline code/modules/food_and_drinks/pizzabox.dm]: use `if(!user.transferItemToLoc(add, src))` instead

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/food_and_drinks/pizzabox.dm [hunk 7]
```diff
@@ -135,22 +153,29 @@
...
 				return
-			else
-				to_chat(user, "<span class='notice'>The stack is dangerously high!</span>")
+			boxes += add
+			newbox.boxes.Cut()
+			newbox.forceMove(src)
```
Comment (Cyberboss) [inline code/modules/food_and_drinks/pizzabox.dm]: And then the forceMove wont be necessary

Comment (Cyberboss) [inline code/modules/food_and_drinks/pizzabox.dm]: And then the forceMove wont be necessary

**Review by Cyberboss — DISMISSED** [review]

**Comment by The-clown-known-as-Pepe** [comment]
sorry i don't speak italics

**Comment by ModDownloading** [comment]
Now its time for me to get a stack of pizza bombs! Cluster pizza grenade (since the boxes have a chance of popping open when thrown)!

**Comment by Bawhoppen** [comment]
good pr accidentally almost shut off my PC to comment on here

**Pull request merged**